### PR TITLE
Features/lose game and restart

### PR DESCRIPTION
**Pull request by shaylasry**

### CODE REVIEW

File: Assets/BowAndArrowImproved/Scripts/BAAIWaveSpawner.cs [hunk 1]
```diff
@@ -18,22 +18,31 @@ public class BAAIWaveSpawner : MonoBehaviour
     private int _currentWave = 1;
     private int _enemiesToKill = 0;
     private List<GameObject> _spawnedEnemies = new();
+    private bool _isLostGame;
```
Comment (IBLabs) [inline Assets/BowAndArrowImproved/Scripts/BAAIWaveSpawner.cs]: this is not proper english, change the name to `_didLostGame`

File: Assets/BowAndArrowImproved/Scripts/BAAIWaveSpawner.cs [hunk 2]
```diff
@@ -85,7 +94,28 @@ private IEnumerator SpawnEnemiesCoroutine(List<GameObject> enemiesToSpawn)
...
+            {
+                enemiesToDestroy.Add(curEnemy);
+            }
+        }
+        
+        foreach (BAAIEnemy enemyToDestroy in enemiesToDestroy)
+        {
+            enemyToDestroy.Die(false);
+        }
+    }
```
Comment (IBLabs) [inline Assets/BowAndArrowImproved/Scripts/BAAIWaveSpawner.cs]: we talked about this

File: Assets/BowAndArrowImproved/Scripts/Enemies/BAAIEnemy.cs [hunk 3]
```diff
@@ -29,15 +29,20 @@ private void OnEnable()
...
 
-        if (playSound)
+        if (deathByArrow)
         {
-            AudioSource.PlayClipAtPoint(breakClips[Random.Range(0, breakClips.Count)], transform.position);    
+            AudioSource.PlayClipAtPoint(breakClips[Random.Range(0, breakClips.Count)], transform.position);
+            UpdateScoreOnDeath scriptComponent = gameObject.GetComponent<UpdateScoreOnDeath>();
+            if (scriptComponent != null)
+            {
+                onDeath.AddListener( scriptComponent.UpdateScoreboard);
```
Comment (IBLabs) [inline Assets/BowAndArrowImproved/Scripts/Enemies/BAAIEnemy.cs]: there is extra space here after the parenthesis, delete it

File: Assets/BowAndArrowImproved/Scripts/GameManager.cs [hunk 4]
```diff
@@ -1,4 +1,5 @@
 using System;
+using UnityEditor;
```
Comment (IBLabs) [inline Assets/BowAndArrowImproved/Scripts/GameManager.cs]: why did you add this? this will create errors during build

File: Assets/BowAndArrowImproved/Scripts/Ballon.cs [hunk 5]
```diff
@@ -50,6 +50,8 @@ private IEnumerator DieCoroutine(float delay)
 
     private void OnCollisionEnter(Collision other)
     {
-        Die(0);
+        if (other.gameObject.CompareTag("Arrow")){
```
Comment (IBLabs) [inline Assets/BowAndArrowImproved/Scripts/Ballon.cs]: create a [SerializeField] for the tag and make it configurable through the editor

File: Assets/BowAndArrowImproved/Scripts/Enemies/BAAIEnemy.cs [hunk 6]
```diff
@@ -29,15 +29,20 @@ private void OnEnable()
         PlayWarCry();
     }
 
-    public void Die(bool playSound)
+    public void Die(bool deathByArrow)
```
Comment (IBLabs) [inline Assets/BowAndArrowImproved/Scripts/Enemies/BAAIEnemy.cs]: change name to `killedByPlayer`
what is tomorrow it's a bomb? or a gun? then you'll need to rename this

File: Assets/BowAndArrowImproved/Scripts/Enemies/BAAIEnemy.cs [hunk 7]
```diff
@@ -29,15 +29,20 @@ private void OnEnable()
...
     {
         gameObject.SetActive(false);
 
-        if (playSound)
+        if (deathByArrow)
         {
-            AudioSource.PlayClipAtPoint(breakClips[Random.Range(0, breakClips.Count)], transform.position);    
+            AudioSource.PlayClipAtPoint(breakClips[Random.Range(0, breakClips.Count)], transform.position);
+            UpdateScoreOnDeath scriptComponent = gameObject.GetComponent<UpdateScoreOnDeath>();
```
Comment (IBLabs) [inline Assets/BowAndArrowImproved/Scripts/Enemies/BAAIEnemy.cs]: wtf is the meaning of the name `scriptComponent`?

File: Assets/BowAndArrowImproved/Scripts/Enemies/BAAIEnemy.cs (hunk 3 above)
Comment (IBLabs) [inline Assets/BowAndArrowImproved/Scripts/Enemies/BAAIEnemy.cs]: you are adding a listener for `onDeath` right before calling `Die()`? wtf is this? what are you trying to do? why not listen to the event like a normal person when the objects is started? very weird...

File: Assets/BowAndArrowImproved/Scripts/Enemies/BAAIEnemy.cs [hunk 8]
```diff
@@ -48,7 +53,9 @@ private void PlayWarCry()
 
     private void OnCollisionEnter(Collision other)
     {
-        Die(true);
+        if (other.gameObject.CompareTag("Arrow")){
```
Comment (IBLabs) [inline Assets/BowAndArrowImproved/Scripts/Enemies/BAAIEnemy.cs]: again make the tag configurable

File: Assets/BowAndArrowImproved/Scripts/HealthBar.cs [hunk 9]
```diff
@@ -9,7 +9,8 @@ public class HealthBar : MonoBehaviour
     [SerializeField] private GameObject healthSquare;
     
     [SerializeField] private HitCountHittable hittable;
-
+    [SerializeField] private int health;
```
Comment (IBLabs) [inline Assets/BowAndArrowImproved/Scripts/HealthBar.cs]: this is a UI class, it only represents the UI of the health bar, it shouldn't hold the health count, only HitCountHittable, no one else!

File: Assets/BowAndArrowImproved/Scripts/HealthBar.cs [hunk 10]
```diff
@@ -27,9 +28,12 @@ public void OnTookHit()
         Destroy(lastSquare);
     }
 
-    private void SetupHealthBar()
+    public void SetupHealthBar()
     {
-        for (int i = 0; i < hittable.health; i++)
+        hittable.SetCount(health);
```
Comment (IBLabs) [inline Assets/BowAndArrowImproved/Scripts/HealthBar.cs]: UI classes DO NOT set things like health, they only consume it to show the UI, not good

File: Assets/BowAndArrowImproved/Scripts/HitCountHittable.cs [hunk 11]
```diff
@@ -1,20 +1,21 @@
...
-
+    
     public UnityEvent onDeath;
     public UnityEvent onHit;
 
+    public void SetCount(int baseHitCount)
+    {
+        hitCount = baseHitCount;
+    }
+
```
Comment (IBLabs) [inline Assets/BowAndArrowImproved/Scripts/HitCountHittable.cs]: we're only setting health from editor right now, no reason to do it from code, remove this function

File: Assets/BowAndArrowImproved/Scripts/UpdateScoreOnDeath.cs [hunk 12]
```diff
@@ -13,8 +13,8 @@ public void Start()
         }
     }
 
-    public void UpdateScoreboard()
+    public void UpdateScoreboard(GameObject arg0)
```
Comment (IBLabs) [inline Assets/BowAndArrowImproved/Scripts/UpdateScoreOnDeath.cs]: why do you need the extra argument here?

File: Assets/BowAndArrowImproved/Scripts/UpdateScoreOnDeath.cs [hunk 13]
```diff
@@ -13,8 +13,8 @@ public void Start()
...
     {
-        Scoreboard.IncreaseScore();
+        Scoreboard.IncreaseScore(1);
```
Comment (IBLabs) [inline Assets/BowAndArrowImproved/Scripts/UpdateScoreOnDeath.cs]: no "magic numbers", make this number configurable through the editor

File: Assets/BowAndArrowImproved/Scripts/WaveStartHandler.cs [hunk 14]
```diff
@@ -1,30 +1,49 @@
 using Unity.Mathematics;
 using UnityEngine;
 using UnityEngine.Events;
+using UnityEngine.Serialization;
```
Comment (IBLabs) [inline Assets/BowAndArrowImproved/Scripts/WaveStartHandler.cs]: why do we need this here? if there is no need, please remove

**Review by IBLabs — CHANGES_REQUESTED** [review]

File: Assets/BowAndArrowImproved/Scripts/UpdateScoreOnDeath.cs (hunk 12 above)
Comment (shaylasry) [reply Assets/BowAndArrowImproved/Scripts/UpdateScoreOnDeath.cs:16]: must use it so I can add a listener to this method from code, changed the name to enemyToAddListener

Comment (shaylasry) [reply Assets/BowAndArrowImproved/Scripts/UpdateScoreOnDeath.cs]: needed it so I could add a listener to the relevant enemy game object from the code
**changed implementation and it's been deleted.**

File: Assets/BowAndArrowImproved/Scripts/Enemies/BAAIEnemy.cs (hunk 3 above)
Comment (shaylasry) [reply Assets/BowAndArrowImproved/Scripts/Enemies/BAAIEnemy.cs]: Removed it and use a boolean in script instead to check if update of score is valid - it will only be valid if the enemy got hit by attack

File: Assets/BowAndArrowImproved/Scripts/Enemies/BAAIEnemy.cs [hunk 15]
```diff
@@ -29,15 +32,20 @@ private void OnEnable()
...
-        if (playSound)
+        if (killedByPlayer)
         {
-            AudioSource.PlayClipAtPoint(breakClips[Random.Range(0, breakClips.Count)], transform.position);    
+            AudioSource.PlayClipAtPoint(breakClips[Random.Range(0, breakClips.Count)], transform.position);
+            UpdateScoreOnDeath updateScoreOnDeathScript = gameObject.GetComponent<UpdateScoreOnDeath>();
+            if (updateScoreOnDeathScript != null)
+            {
+                updateScoreOnDeathScript.isUpdateValid = true;
+            }
```
Comment (IBLabs) [inline Assets/BowAndArrowImproved/Scripts/Enemies/BAAIEnemy.cs]: bad code, nobody cares about this script here - pass another argument to the event

**Review by IBLabs — APPROVED** [review]

**Review by IBLabs — APPROVED** [review]
